### PR TITLE
Merge resolved_paths and watched_paths into additional_paths

### PR DESCRIPTION
**Pull request by CGA1123**
Merges the configuration of resolved_paths and watched_paths together
into additional_paths, while keeping backwards compatibility for both.

watched_paths and resolved_paths are deprecated in favour of
additional_paths.

### CODE REVIEW

**Comment by gauravtiwari** [comment]
@CGA1123 Thanks for working on this. 

I have been thinking to actually consolidate `watched` and `resolved` paths since they are pretty much same and server similar purpose. Perhaps, we can add `additional_paths` that covers both and remove both `watched` and `resolved` option from the settings?

**Comment by CGA1123** [comment]
@jakeNiemiec `bower_components` is the example used in the documentation https://github.com/rails/webpacker/blob/740b686dd9ba5e97dbc679aa94e4091c0ca8415c/lib/webpacker/compiler.rb#L6

@gauravtiwari That sounds good 😄Should this be done in a backwards-compatible way? Or is it okay to fully remove without deprecation of those options?

**Comment by gauravtiwari** [comment]
Thanks @CGA1123 Lets do it backwards-compatible way in this PR, please.

**Comment by CGA1123** [comment]
@gauravtiwari One difference between `resolved_paths` and `watched_paths` is that `resolved_paths` is actually passed to `webpack` in order to resolve module names, while `watched_paths` is used only by `webpacker` to determine whether there is a need to recompile.

Is it okay to merge those _slightly_ different concerns? Is there any issue with passing on those extra 'watched' paths to webpack?

**Comment by gauravtiwari** [comment]
Yes that's correct. 

I guess whatever modules are resolved or watched should be available to all. I think that's what you are trying to solve in this pr, being resolved paths to compiler? 

So, if we just use one setting to add additional paths that can be used by - webpack,  on-demand compiler and sass loader then it would remove complexity and confusion.

**Comment by gauravtiwari** [comment]
Am I missing something?

**Comment by CGA1123** [comment]
@gauravtiwari No, that makes sense.

**Comment by gauravtiwari** [comment]
Great, thank you 👍 

Can we also add deprecation warnings to old options?

**Comment by CGA1123** [comment]
I've added warnings to `resolved_paths` and `watched_paths` which will emit when they are non-empty.

File: lib/webpacker/configuration.rb [hunk 1]
```diff
@@ -76,6 +76,14 @@ def extract_css?
   end
 
   private
+    def resolved_paths
+      resolved = data.fetch(:resolved_paths, [])
+
+      warn "The resolved_paths option has been deprecated. Use additional_paths instead." unless resolved.empty?
+
+      resoved
```
Comment (gauravtiwari) [inline lib/webpacker/configuration.rb:84]: typo here ^^ 

Maybe lets call it paths instead of `resolved`?

**Comment by gauravtiwari** [comment]
Thanks @CGA1123, looking great. I will review this later today and get it merged.

**Comment by CGA1123** [comment]
Sound good @gauravtiwari 😄Thanks for your time!

**Comment by cmitz** [comment]
👋🏽  Can someone please put this _somewhere_ into a changelog as breaking change?

**Comment by CGA1123** [comment]
Hi @cmitz 

When initially implemented this was done in a backwards-compatible manner.

It seems more recently in e81793ba3742a1e33176e7796f2b8586c73712b3 / #2802 backwards compatibility was removed in preparation with the release of the new major version 6.

I don't see any reason a PR to add this to the `CHANGELOG` or `6_0_upgrade` would not be accepted before moving the 6.0.0 beta release into a full release 😄

I am assuming that you are running the `6.x` beta release, as the `5.x` release line should maintain backwards compatibility.

**Comment by cmitz** [comment]
I encountered this when upgrading from 5.1.1 to 5.2.1. We did not have a `additional_paths` entry in `webpacker.yml` yet, resulting in the log:

```
NoMethodError: undefined method `+' for nil:NilClass
/srv/.../shared/bundle/ruby/2.7.0/gems/webpacker-5.2.1/lib/webpacker/configuration.rb:31:in `additional_paths'
...
```

I'm happy to create a PR to add it to the Changelog for 5.2.0, but if you say "was done in a backwards-compatible manner" I guess a new issue is more appropriate?

**Comment by CGA1123** [comment]
@cmitz Yes, in this case, that is a bug and not intended behaviour (my bad). I think a new issue would be worth it 😄 

That is an interesting `nil` issue 🤔 
The `additional_paths` should be falling back to defaults set in the YAML afaiu:
https://github.com/rails/webpacker/blob/3d73dbad1eeaf3ef1b7aeaba4ba9262d507f999e/lib/webpacker/configuration.rb#L30-L32
https://github.com/rails/webpacker/blob/3d73dbad1eeaf3ef1b7aeaba4ba9262d507f999e/lib/webpacker/configuration.rb#L87-L89
https://github.com/rails/webpacker/blob/3d73dbad1eeaf3ef1b7aeaba4ba9262d507f999e/lib/webpacker/configuration.rb#L109-L112
https://github.com/rails/webpacker/blob/3d73dbad1eeaf3ef1b7aeaba4ba9262d507f999e/lib/install/config/webpacker.yml#L11-L13

**Comment by cmitz** [comment]
Thanks for clearing that up 😄 , I've made an issue for it: https://github.com/rails/webpacker/issues/2866

**Pull request merged**